src/Microsoft.Framework.FileSystemGlobbing/Internal/PatternContexts/PatternContextLinear.cs

```diff
@@ -30,6 +31,14 @@ public override void PushDirectory(DirectoryInfoBase directory)
             }
             else
             {
+                // Determine this frame's contribution to the stem (if any)
+                var segment = Pattern.Segments[Frame.SegmentIndex];
+                if (Frame.InStem || segment.ProducesStem)
+                {
+                    Frame.InStem = true;
+                    Frame.Stem = MatcherContext.CombinePath(Frame.Stem, directory.Name);
```

Comment:
I'm changing it to be a list of strings that will be joined together at the end. This, plus making the FilePatternMatch object a struct means the only additional allocation is the final stem string. I considered making that lazy, but to do that you'd have to hold all the individual stem items, which would basically take up the same space.

I'll do some before/after memory tracing to confirm.